src/ring_buffer.c

```diff
@@ -0,0 +1,83 @@
+/**
+ * @Author: S. Sharma <silentcat>
+ * @Date:   2019-03-07T10:32:56-06:00
+ * @Email:  silentcat@protonmail.com
+ * @Last modified by:   silentcat
+ * @Last modified time: 2019-03-07T22:58:26-06:00
+ */
+
+#include "include/ring_buffer.h"
+#include <stdio.h>
+
+enum cc_stat rbuf_new(Rbuf **rbuf)
+{
+    RbufConf conf;
+    rbuf_conf_init(&conf);
+    return rbuf_conf_new(&conf, rbuf);
+}
+
+enum cc_stat rbuf_conf_new(RbufConf *rconf, Rbuf **rbuf)
+{
+    Rbuf *ringbuf = rconf->mem_alloc(sizeof(Rbuf));
+    if (!ringbuf)
+       return CC_ERR_ALLOC;
+    if (!(ringbuf->buf = rconf->mem_calloc(DEFAULT_RBUF_CAPACITY, sizeof(Rbuf)))) {
+      rconf->mem_free(ringbuf);
+      return CC_ERR_ALLOC;
+    }
+
+    for (int i = 0; i < DEFAULT_RBUF_CAPACITY; i++)
+        ringbuf->buf[i] = 0x0;
+
+    ringbuf->mem_alloc = rconf->mem_alloc;
+    ringbuf->mem_calloc = rconf->mem_calloc;
+    ringbuf->mem_free = rconf->mem_free;
+    ringbuf->capacity = DEFAULT_RBUF_CAPACITY;
+    ringbuf->size = 0;
+    ringbuf->head = 0, ringbuf->tail = 0;
+    *rbuf = ringbuf;
+    return CC_OK;
+}
+
+enum cc_stat rbuf_conf_init(RbufConf *rconf)
+{
+    rconf->capacity = DEFAULT_RBUF_CAPACITY;
+    rconf->mem_alloc = malloc;
+    rconf->mem_calloc = calloc;
+    rconf->mem_free = free;
+    return CC_OK;
+}
+
+void rbuf_destroy(Rbuf *rbuf)
+{
+    rbuf->mem_free(rbuf->buf);
+    rbuf->mem_free(rbuf);
+}
+
+int rbuf_is_empty(Rbuf *rbuf)
+{
+    return (rbuf->size == 0);
+}
+
+enum cc_stat rbuf_enqueue(Rbuf *rbuf, uint64_t item)
+{
+    if (!rbuf)
+       return CC_ERR_ALLOC;
+    rbuf->buf[rbuf->head] = item;
+    rbuf->head = (rbuf->head + 1) % rbuf->capacity;
```

Comment:
Let's say the `capacity` is 10 and I enqueue 12 things in a row. Is the `tail` still valid? 

When the head wraps around and catches up with the tail, shouldn't it start pushing the tail in front of it every time it moves forward?